### PR TITLE
unlink memory shared file immediately on linux

### PR DESCRIPTION
**Pull request by piiswrong**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by piiswrong** [comment]
I can't reproduce.
Maybe this would help?
```
        if sys.version_info[0] == 2:
            new_fd = multiprocessing.reduction.rebuild_handle(fd)
        else:
            new_fd = fd.detach()
        ret = nd.NDArray(nd.ndarray._new_from_shared_mem(pid, new_fd, shape, dtype))
        return ret
```

**Comment by Vikas89** [comment]
Thanks for taking a look.
Let me try to give exact repro steps on ec2 instance - 

1. Spin up p3.2xl
2. which python
/home/ubuntu/anaconda3/bin//python
3. /home/ubuntu/anaconda3/bin/pip install mxnet 
4. copy test code  to test.py
```
from multiprocessing import Queue, Event, Process
import mxnet as mx
import time

class Producer(object):
    
    def __init__(self, input_queue):
        self.input_queue = input_queue
        
    def fill(self, length=100):
        for i in range(length):
            self.input_queue.put(mx.nd.ones((1,1)))

input_queue = Queue()
producer = Producer(input_queue)
# producer.fill()
preprocessor_process = Process(target=producer.fill)
preprocessor_process.daemon = True
preprocessor_process.start()
# preprocessor_process2 = Process(target=producer.fill)
# preprocessor_process2.daemon = True
# preprocessor_process2.start()

# Read without new thread
while True:
    print(input_queue.get())

class Consumer(object):
    
    def __init__(self, input_queue):
        self.input_queue = input_queue
        
    def read(self):
        while True:
            print(self.input_queue.get().shape)

consumer = Consumer(input_queue)
consumer_process = Process(target=consumer.read)
consumer_process.daemon = True
consumer_process.start()

time.sleep(1000000)
```
5. python test.py

Meanwhile if I do /home/ubuntu/anaconda3/bin/pip install mxnet==1.2 
There is no error and it works well.

**Comment by Roshrini** [comment]
Also, this error can be reproduced with Python3.6.. With Python2.7 test code runs fine.

**Comment by Vikas89** [comment]
@piiswrong I tried your suggested change - 
```
  if sys.version_info[0] == 2:
            new_fd = multiprocessing.reduction.rebuild_handle(fd)
        else:
            new_fd = fd.detach()
        ret = nd.NDArray(nd.ndarray._new_from_shared_mem(pid, new_fd, shape, dtype))
        return ret
```

the doesn't fix the issue.

**Comment by Roshrini** [comment]
Also tried that fix, didn't work

**Comment by Vikas89** [comment]
other related issues: https://github.com/apache/incubator-mxnet/issues/13126

@piiswrong  Can you explain what this PR is trying to achieve and since it causes regression one possibility is  may be need to revert this. Are you aware of any possible consequences of reverting this PR?

**Comment by kaonashi-tyc** [comment]
Watching. Can you share some consideration why this feature is removed in the first place?

**Comment by Vikas89** [comment]
@piiswrong  can you explain more about this change and the intent behind the change. There is not enough information in details section and it would be useful if you can tell the motivation behind doing this.

For this particular case: https://github.com/apache/incubator-mxnet/pull/10992#issuecomment-451534918
I have a possible explanation and recommended way to fix: 
We saw that if  process doesn't die, we are able to access the ndarray without any trouble.I don't know how multiprocessing queue in python works, but SO link told me that put doesn't guarantee that objects are pickled and in the queue. If I see the stack of error, it comes in consumer which is trying to access a file which was created by Producer process. This means that pickling is happening I am getting from queue. Since process which created that memory already died, the memory is gone, user is getting error because consumer is trying to pickle from a memory which is gone. 

Recommended fix:
1) Fix the code as recommended, don't let your producer die until the consumers has consumed.
2) explicitly pickle ndarray into queue, (see SO link I shared) , this will ensure that objects are indeed placed in the queue from mxnet memory of the process and removes any un-certainity of when objects are getting pickled(before the process dies or after the process dies)

**Comment by szha** [comment]
@Vikas89 @Roshrini @kaonashi-tyc I recommend tracking the problem in a new issue. Could one of you create one and link it here? Thanks.
